### DISCUSSION TITLE
API endpoints similar to OpenAI

### DISCUSSION THREAD

**Discussion by bloodyburger** [open-webui/open-webui#2252] [General]
Does this setup offer endpoints for Chat completion and embeddings similar to what OpenAI is offering? I am exploring to use the self hosted models via APIs to build a custom chatbot.

**Comment by Kisaragi-ng** [comment]
I'm confused with this too, I thought open-webui already have openai-compatible api. trying to search this using search engine is pure pain since it keeps pointing me https://docs.openwebui.com/tutorial/openai

anyway, this is in the readme;
```
🔑 API Key Generation Support: Generate secret keys to leverage Open WebUI with OpenAI libraries, simplifying integration and development.
```

but when i created an api key in Settings -> Account, it doesn't seems i can use this properly.
the very basic command to use in https://platform.openai.com/docs/api-reference is this:

```bash
curl https://api.openai.com/v1/models \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -H "OpenAI-Organization: YOUR_ORG_ID" \
  -H "OpenAI-Project: $PROJECT_ID"
```

but since open-webui has no configuration for `-Organization` and `-Project`, I altered the curl command like this:

```bash
curl https://example.de/v1/models -H "Authorization: Bearer sk-9b30b****"
```
the output:
```html
<!DOCTYPE html>
<html lang="en">
        <head>
                <meta charset="utf-8" />
                <link rel="icon" href="/favicon.png" />
                <link rel="manifest" href="/manifest.json" crossorigin="use-credentials" />
                <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />
                <meta name="robots" content="noindex,nofollow" />
                <link
                        rel="search"
                        type="application/opensearchdescription+xml"
                        title="Open WebUI"
                        href="/opensearch.xml"
                />
                <script>
                        // On page load or when changing themes, best to add inline in `head` to avoid FOUC
...snip...
                                style="
                                        position: absolute;
                                        width: 6rem;
                                        height: 6rem;
                                        top: 46%;
                                        left: 50%;
                                        margin: -40px 0 0 -40px;
                                "
                                src="/logo.svg"
                        />

                        <!-- <span style="position: absolute; bottom: 32px; left: 50%; margin: -36px 0 0 -36px">
                                Footer content
                        </span> -->
                </div>
        </body>
</html>
```

this output probably just render the Open WebUI logo using it's default template css.

I'm trying to figure this out since i really want to use my open-webui instance in Obsidian, with this plugin https://github.com/pfrankov/obsidian-local-gpt 

![obsidian](https://github.com/open-webui/open-webui/assets/88741696/cc4855d4-b5d9-4943-930a-48d54dd231af)

but i guess it's not possible, and i have to connect this plugin directly to ollama, making me have to expose it's port to my lan / tailscale network:

![ollama](https://github.com/open-webui/open-webui/assets/88741696/e21b161e-c60f-4640-9d0d-e2f11991726b)

---

after about half an hour trying to write this reply I just realized I might be confusing open-webui with https://lmstudio.ai/docs/local-server 
since they have this:
```
# Supported endpoints

GET /v1/models
POST /v1/chat/completions
POST /v1/embeddings            
POST /v1/completions
```
I need some sleep. so I guess I'll be using LM Studio or illama.cpp for awhile until open-webui have similar supported endpoints as I needed API access more than webui for time being.

---
edit, I finally managed to get it work.

apparently I have to append `/ollama` in my base url, so it will looks like this:
```
https://example.de/ollama
```
it can fetch the model list just fine.

![slash ollama](https://github.com/open-webui/open-webui/assets/88741696/3c12ebf0-ea2b-4c5b-8695-c2fa90e6ec7e)

here i made a short recording while it's running. It's pretty wonky but it prove that it'll works.

https://github.com/open-webui/open-webui/assets/88741696/9dd7f679-91a4-478d-aeb1-d9c75d608c5e

my open-webui were behind cloudflared tunnel, so it's good to see that it'll works.

---
edit(1): adding screenshot
edit(2): it's sorta work
